Feature/cart

SwanyCastle opened the pull request:
### Pull Request

> ### 제목
>
> - Feat: CartException.java 생성 및 GlobalExceptionHandler.java 에 추가
> - Feat: Cart, CartItem Entity 생성 및 CartDto, CartItemDto 생성
> - Feat: 회원 가입 및 소셜 로그인시 장바구니 생성
> - Feat: 특정 유저의 장바구니 목록 조회
> - Fix: Table 'ecommerce.cart_item' doesn't exist 에러 해결
> - Feat: 장바구니에 상품 담기 기능 구현
> - Style: 코드 스타일 정렬
> - Fix: 장바구니에 상품 담기 기능 수정
> - Refactor: 사용하지 않는 import 제거
> - Feat: 장바구니 상품 수량 수정
> - Feat: 장바구니 상품 삭제
> - Feat: 장바구니 상품 전체 제거
> - Refactor: 장바구니 목록 조회, 장바구니 상품 추가, 수정, 삭제, 전체 삭제 API End Point URI 변경

> ### 변경사항
> 
> 📌 **AS-IS**
> - Builder.Default 를 활용해 ProductStatus 초기값 설정
> - 상품 등록, 조회, 수정, 삭제 테스트 코드 작성
>
> 🔖 **TO-BE**
>  🔧 장바구니 목록 조회, 장바구니 상품 추가, 수정, 삭제, 전체삭제 기능 구현
> - Cart, CartItem Entity 생성 및 CartDto, CartItemDto 생성
>     - CartDto 의 fromEntity 에서 cartItems 요소들의 가격의 총합을 계산해서 total_price 필드로 응답
>     - CartItemDto 의 fromEntity 에서 상품 가격에 장바구니의 상품 갯수 만큼의 합산을 price 필드로 응답
>
> - 회원 가입 및 소셜 로그인시 장바구니 생성
>     - 일반 회원 회원가입시 회원 권한이 CUSTOMER 인 회원만 장바구니 생성
>     - 소셜 로그인시 Cart 테이블에 소셜 로그인한 사용자의 ID 를 가지는 장바구니가 없다면 생성
>
> - Table 'ecommerce.cart_item' doesn't exist 에러 해결
>      - Table 이름이 CartItem 처럼 카멜 케이스 일 때 hibernate 의 기본 매핑 규칙을 따라
>          hibernate 가 자동으로 카멜 케이스 -> 스네이크 케이스 (cartItem -> cart_item) 로
>          변경해서 CartItem 테이블이 아니라 cart_item 테이블을 참조하려는 에러가 발생
>
>      - 해결방법
>          1. 테이블 이름을 CartItem -> Cart_Item 으로 변경
>          2. application.yml 에
>              spring:
>                jpa:
>                  hibernate:
>                      naimg:
>                         physical-strategy:   org.hibernate.boot.model.naming.PhysicalNamingStrategyStandardImpl 추가
>
> - 장바구니에 상품 담기 기능 구현
>    - PreAuthorize 어노테이션으로 회원 권한이 CUSTOMER 인 유저만 요청 가능하도록 설정
>    - Path Variable 로 받는 memberId 와 token 에 있는 memberId 가 동일한지 확인
>    - memberId 에 해당하는 cart 객체와 request 로 받은 productId 에 해당하는 product 객체 조호후
>      product 객체가 활성화 되어있는지 상태를 확인 하고 cart, product 객체를 함께 가지는 cartItem 이 있는지 확인
>    - 위의 조건이 통과되면 데이터 저장하고 성공 응답
>
> - 장바구니에 상품 담기 기능 수정
>    - 장바구니에 상품 단기 성공후 응답 ResponseDto 에서 CartItemDto.Response 로 변경
>
> - 장바구니 상품 수량 수정
>    - Path Variable 로 받은 memberId 와 token 에 있는 memberId 동일한지 확인
>    - cartItemId 를 가지는 cartItem 이 존재하는지 확인
>    - 수정 하려는 장바구니 상품 수량이 상품의 재고보다 큰지 확인후 크지 않다면 수정하여 응답 구현
>
> - 장바구니 상품 삭제
>    - Path Variable 로 받은 memberId 와 token 에 있는 memberId 일치 확인
>    - cartItemId 에 해당하는 장바구니 상품이 있는지 확인 후 있다면 삭제하고 성공 응답
>
> - 장바구니 상품 전체 제거
>    - Path Variable 로 받은 memberId 와 token 에 있는 memberId 일치 확인
>    - memberId 에 해당하는 cart 객체를 포함하는 모든 cartItem 데이터 삭제후 성공 응답

> ### 테스트
>
> - [ ] 테스트 코드
> - [x] API 테스트

SwanyCastle committed:
[Feat]: CartException.java 생성 및 GlobalExceptionHandler.java 에 추가

SwanyCastle committed:
[Feat]: Cart, CartItem Entity 생성 및 CartDto, CartItemDto 생성

SwanyCastle committed:
[Feat]: 회원 가입 및 소셜 로그인시 장바구니 생성

회원 가입 및 소셜 로그인시 장바구니 생성
- 일반 회원 회원가입시 회원 권한이 CUSTOMER 인 회원만 장바구니 생성
- 소셜 로그인시 Cart 테이블에 소셜 로그인한 사용자의 ID 를 가지는 장바구니가 없다면 생성

SwanyCastle committed:
[Feat]: 특정 유저의 장바구니 목록 조회

SwanyCastle committed:
[Fix]: Table 'ecommerce.cart_item' doesn't exist 에러 해결

Table 'ecommerce.cart_item' doesn't exist 에러 해결
- Table 이름이 CartItem 처럼 카멜 케이스 일 때 hibernate 의 기본 매핑 규칙을 따라
  hibernate 가 자동으로 카멜 케이스 -> 스네이크 케이스 (cartItem -> cart_item) 로
  변경해서 CartItem 테이블이 아니라 cart_item 테이블을 참조하려는 에러가 발생
- 해결방법
  1. 테이블 이름을 CartItem -> Cart_Item 으로 변경
  2. application.yml 에
     spring:
       jpa:
         hibernate:
           naimg:
             physical-strategy:   org.hibernate.boot.model.naming.PhysicalNamingStrategyStandardImpl 추가

SwanyCastle committed:
[Feat]: 장바구니에 상품 담기 기능 구현

장바구니에 상품 담기 기능 구현
- PreAuthorize 어노테이션으로 회원 권한이 CUSTOMER 인 유저만 요청 가능하도록 설정
- Path Variable 로 받는 memberId 와 token 에 있는 memberId 가 동일한지 확인
- memberId 에 해당하는 cart 객체와 request 로 받은 productId 에 해당하는 product 객체 조호후
  product 객체가 활성화 되어있는지 상태를 확인 하고 cart, product 객체를 함께 가지는 cartItem 이 있는지 확인
- 위의 조건이 통과되면 데이터 저장하고 성공 응답

SwanyCastle committed:
[Style]: 코드 스타일 정렬

SwanyCastle committed:
[Fix]: 장바구니에 상품 담기 기능 수정

장바구니에 상품 담기 기능 수정
- 장바구니에 상품 단기 성공후 응답 ResponseDto 에서 CartItemDto.Response 로 변경

SwanyCastle committed:
[Refactor]: 사용하지 않는 import 제거

SwanyCastle committed:
[Feat]: 장바구니 상품 수량 수정

장바구니 상품 수량 수정
- Path Variable 로 받은 memberId 와 token 에 있는 memberId 동일한지 확인
- cartItemId 를 가지는 cartItem 이 존재하는지 확인
- 수정 하려는 장바구니 상품 수량이 상품의 재고보다 큰지 확인후 크지 않다면 수정하여 응답 구현

SwanyCastle committed:
[Feat]: 장바구니 상품 삭제

장바구니 상품 삭제
- Path Variable 로 받은 memberId 와 token 에 있는 memberId 일치 확인
- cartItemId 에 해당하는 장바구니 상품이 있는지 확인 후 있다면 삭제하고 성공 응답

SwanyCastle committed:
[Feat]: 장바구니 상품 전체 제거

장바구니 상품 전체 제거
- Path Variable 로 받은 memberId 와 token 에 있는 memberId 일치 확인
- memberId 에 해당하는 cart 객체를 포함하는 모든 cartItem 데이터 삭제후 성공 응답

SwanyCastle committed:
[Refactor]: 장바구니 목록 조회, 장바구니 상품 추가, 수정, 삭제, 전체 삭제 API End Point URI 변경

hjkim22 commented:
장바구니 관련 기능 잘봤습니다. 테이블 문제도 해결하셨네요!
yml에서 PhysicalNamingStrategyStandardImpl 설정하고 매핑 문제를 해결한 방식도 좋은거같아요 Hibernate 기본 규칙을 맞추는 과정이 중요하다는 것을 느꼈습니다.
코드 잘봤습니다 ㅎㅎ

spring-kotlin-lover commented on src/main/java/com/ecommerce/service/cart/CartItemServiceImplement.java:
```diff
@@ -0,0 +1,175 @@
+    boolean isExists = cartItemRepository.existsByCartAndProduct(cart, product);
+    if (isExists) {
+      throw new CartException(ResponseCode.CART_ITEM_ALREADY_EXISTS);
+    }
```
이미 cartItem 에 해당하는 상품이 담겨있는 경우, 예외를 던지지말고, 요청된 수량만큼 더해주는 방향은 어떨까요? 왜냐하면 상품상세에서 cart 담기를 했을 때 이미 들어있는 상품이라면 수량을 더해주는 유저 flow 가 더 사용성이 좋기 때문입니다.

spring-kotlin-lover commented on src/main/java/com/ecommerce/service/cart/CartItemServiceImplement.java:
```diff
@@ -0,0 +1,175 @@
+   * @return Cart
+   */
+  @Transactional(readOnly = true)
+  public Cart getCartByMemberId(String memberId) {
```
Cart 를 조회하는 로직은 CartService 에 선언하는게 맞지 않을까요? CartItemService 와 CartService 의 역할을 한번 생각해보시면 좋을 것 같습니다~!

SwanyCastle committed:
[Fix]: PR 리뷰 코멘트 사항 수정 및 테스트코드 작성

PR 리뷰 코멘트 사항 수정 및 테스트코드 작성
- CartItem 에 상품이 담겨있는 경우 예외 던지지 않고, 요청된 수량 만큼 더해주도록 수정
  - 장바구니에 상품 추가시 요청된 수량과 상품의 재고에 대한 예외처리 추가
  - 장바구니 상품 수량 수정시에도 동일하게 요청된 수량과 상품의 재고에 대한 예외처리 수정
- Cart 를 조회하는 getCartByMemberId 메서드를 CartService 로 이동
- 장바구니에 상품 담기, 수정 삭제, 전체 삭제 기능에 대한 테스트 코드 작성

SwanyCastle committed:
[Fix]: Member, Product 정보 삭제시 테스트 코드 추가

Member, Product 정보 삭제시 테스트 코드 추가
- memberRepository, productRepository 의 삭제 메서드에 대한 verify() 테스트 추가

SwanyCastle committed:
[Refactor]: ResponseCode 상수 앞에 도메인 이름 표시
